### PR TITLE
Google ads enhanced conversion new cloud mode destination onboarding

### PR DESCRIPTION
**Pull request by ItsSudip**
## Description of the change

> Description here

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: v0/destinations/google_ads_enhanced_conversion/transform.js [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+  if (typeof payload.partialFailure !== "boolean") {
+    throw new CustomError(
+      `[Google_ads_enhanced_marketing]:: partialFailure should be a boolean.`,
+      400
+    );
+  }
+  payload.conversionAdjustments[0].adjustmentType = "ENHANCEMENT";
+  // Removing the null values from userIdentifier
+  const arr = payload.conversionAdjustments[0].userIdentifiers;
+  payload.conversionAdjustments[0].userIdentifiers = arr.filter(function() {
```
Comment (arajguha) [inline v0/destinations/google_ads_enhanced_conversion/transform.js:67]: lets use arrow function here

File: v0/destinations/google_ads_enhanced_conversion/transform.js [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+    return true;
+  });
+  return responseBuilder(metadata, message, destination, payload);
+};
+
+const processEvent = async (metadata, message, destination) => {
+  const { type } = message;
```
Comment (arajguha) [inline v0/destinations/google_ads_enhanced_conversion/transform.js:74]: maybe we can add one more check if `type` actually exists otherwise `.toLowerCase()` on undefined will throw some different error

File: v0/destinations/google_ads_enhanced_conversion/transform.js [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+      throw new CustomError(
+        `[Google_ads_enhanced_marketing]:: loginCustomerId is required as subAccount is true.`,
+        400
+      );
+  return response;
+};
+
+const processTrackEvent = async (metadata, message, destination) => {
+  const payload = constructPayload(message, trackMapping);
+  if (!payload.conversionAdjustments[0].userIdentifiers) {
```
Comment (arajguha) [inline v0/destinations/google_ads_enhanced_conversion/transform.js:52]: should we put a check if `conversionAdjustments` exists on `payload` and if it is of type `Array` etc ?

File: v0/destinations/google_ads_enhanced_conversion/util.js [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+  }
+  // else successfully return status, message and original destination response
+  gaecRespHandler(
+    destinationResponse,
+    "during Google_ads_enhanced_conversion response transformation",
+    TRANSFORMER_METRIC.TRANSFORMER_STAGE.RESPONSE_TRANSFORM
+  );
+};
+// eslint-disable-next-line func-names
+class networkHandler {
```
Comment (arajguha) [inline v0/destinations/google_ads_enhanced_conversion/util.js:153]: lets Keep the class name as `NetworkHandler`. Do we not have a common Base Class for this yet? Seems like this is a repititive piece of code

File: v0/destinations/google_ads_enhanced_conversion/util.js [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+};
+
+/**
+ * This function will handle the responses from google ads api
+ * @param {*} destResponse
+ * @param {*} stageMsg
+ */
+const gaecRespHandler = (destResponse, stageMsg) => {
+  const { status, response } = destResponse;
+  throw new ErrorBuilder()
```
Comment (arajguha) [inline v0/destinations/google_ads_enhanced_conversion/util.js:124]: are we planning setting statTags for alerting for this destination?

File: v0/destinations/google_ads_enhanced_conversion/util.js [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+    );
+    return conversionActionIdResponse;
+  }
+
+  const conversionActionId = get(
+    conversionActionIdResponse,
+    "response.data[0].results[0].conversionAction.id"
+  );
+  if (!conversionActionId)
+    throw new ErrorBuilder()
```
Comment (arajguha) [inline v0/destinations/google_ads_enhanced_conversion/util.js:81]: lets wrap this within `{}` helps in debugging

File: v0/destinations/google_ads_enhanced_conversion/util.js [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+      return "";
+  }
+};
+
+/**
+ * This function will handle the responses from google ads api
+ * @param {*} destResponse
+ * @param {*} stageMsg
+ */
+const gaecRespHandler = (destResponse, stageMsg) => {
```
Comment (arajguha) [inline v0/destinations/google_ads_enhanced_conversion/util.js:122]: are we execting this util function to have more re-use in the future? seems like this is being called only in one place and the code inside it is pretty straight forward.

File: v0/destinations/google_ads_enhanced_conversion/transform.js (hunk 3 above)
Comment (ItsSudip) [reply v0/destinations/google_ads_enhanced_conversion/transform.js:52]: We don't need to check this because the payload will always get created in that way.

File: v0/destinations/google_ads_enhanced_conversion/util.js (hunk 4 above)
Comment (ItsSudip) [reply v0/destinations/google_ads_enhanced_conversion/util.js:153]: Putting the class name as that will not work. With the other destination with transformerProxy support let's say market, pardot are using the same name structure.

File: v0/destinations/google_ads_enhanced_conversion/transform.js (hunk 3 above)
Comment (arajguha) [reply v0/destinations/google_ads_enhanced_conversion/transform.js:52]: are we sure on that? asking because I did not see any `required: true` for the fields in the mapping json. what if they are not there?

Comment (arajguha) [reply v0/destinations/google_ads_enhanced_conversion/transform.js:52]: ok. we would probably need to refactor this part

File: v0/destinations/google_ads_enhanced_conversion/util.js (hunk 4 above)
Comment (arajguha) [reply v0/destinations/google_ads_enhanced_conversion/util.js:153]: ok we would probably need to refactor this later then

File: v0/destinations/google_ads_enhanced_conversion/transform.js (hunk 3 above)
Comment (ItsSudip) [reply v0/destinations/google_ads_enhanced_conversion/transform.js:52]: orderId is required as true under conversionAdjustments[0].

File: v0/destinations/google_ads_enhanced_conversion/util.js (hunk 7 above)
Comment (ItsSudip) [reply v0/destinations/google_ads_enhanced_conversion/util.js:122]: Removed it

File: v0/destinations/google_ads_enhanced_conversion/util.js (hunk 5 above)
Comment (ItsSudip) [reply v0/destinations/google_ads_enhanced_conversion/util.js:124]: It is automatically set.

File: v0/destinations/google_ads_enhanced_conversion/util.js [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+};
+
+/**
+ * This function is responsible for collecting the conversionActionId
+ * and calling the enhanced conversion.
+ * data to customer list.
+ * @param {*} request
+ * @returns
+ */
+const gaemProxyRequest = async request => {
```
Comment (saikumarrs) [inline v0/destinations/google_ads_enhanced_conversion/util.js:54]: gaem?

File: v0/destinations/google_ads_enhanced_conversion/util.js (hunk 7 above)
Comment (saikumarrs) [inline v0/destinations/google_ads_enhanced_conversion/util.js:122]: Why "gaec" in the method name when this is already a part of "gaec" destination?

File: v0/util/index.js [hunk 9]
```diff
@@ -712,7 +712,7 @@ const constructPayload = (message, mappingJson, destinationName = null) => {
       if (value || value === 0 || value === false) {
         if (destKey) {
           // set the value only if correct
-          set(payload, destKey, value);
+          _.set(payload, destKey, value);
```
Comment (saikumarrs) [inline v0/util/index.js:715]: What called for this change? We are already importing a set method on Ln 13.

File: adapters/networkHandlerFactory.js [hunk 10]
```diff
@@ -14,12 +14,17 @@ const {
...
+} = require("../v0/destinations/google_ads_enhanced_conversion/util");
+
 const handler = {
   generic: GenericNetworkHandler,
   braze: BrazeNetworkHandler,
   marketo: MarketoNetworkHandler,
   pardot: PardotNetworkHandler,
-  google_adwords_remarketing_lists: GoogleAdwordsRemarketingListNetworkHandler
+  google_adwords_remarketing_lists: GoogleAdwordsRemarketingListNetworkHandler,
+  google_ads_enhanced_conversion: GoogleAdsEnhancedConversion
```
Comment (saikumarrs) [inline adapters/networkHandlerFactory.js:27]: What is the difference b/w the key (destination) names used here vs in features.json?

File: v0/destinations/google_ads_enhanced_conversion/util.js (hunk 8 above)
Comment (ItsSudip) [reply v0/destinations/google_ads_enhanced_conversion/util.js:54]: Removed this.

File: v0/destinations/google_ads_enhanced_conversion/util.js (hunk 7 above)
Comment (ItsSudip) [reply v0/destinations/google_ads_enhanced_conversion/util.js:122]: Removed this.

File: v0/util/index.js (hunk 9 above)
Comment (ItsSudip) [reply v0/util/index.js:715]: I have used some attributes in trackConfig.json i.e. `"destKey":"conversionAdjustments[0].restatementValue.currencyCode"`. Using the set from set-value is not working in this way, whereas set from lodash working.

File: adapters/networkHandlerFactory.js (hunk 10 above)
Comment (ItsSudip) [reply adapters/networkHandlerFactory.js:27]: Names are the same but in lower case here and uppercase in feature.json.

Comment (saikumarrs) [reply adapters/networkHandlerFactory.js:27]: We should have a single source of this data to avoid errors.
CC @sanpj2292 as this is tech debt.

File: __mocks__/data/google_adwords_remarketing_lists/proxy_response.json [hunk 11]
```diff
@@ -59,5 +59,51 @@
...
+    }
+  },
+  "https://googleads.googleapis.com/v10/customers/1234567891/googleAds:searchStream": {
+    "data": [
+      {
+        "results": [
+          {
+            "conversionAction": {
+              "resourceName": "customers/7693729833/conversionActions/848898416",
+              "id": "848898416"
```
Comment (saikumarrs) [inline __mocks__/data/google_adwords_remarketing_lists/proxy_response.json:84]:
```suggestion
  "https://googleads.googleapis.com/v10/customers/1234567891/googleAds:searchStream": {
    "data
      {
        "results": [
          {
            "conversionAction": {
              "resourceName": "customers/7693729833/conversionActions/848898416",
              "id": "848898416"
```

File: v0/destinations/google_ads_enhanced_conversion/data/trackConfig.json [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+    "destKey": "conversionAdjustments[0].gclidDateTimePair.gclid",
+    "sourceKeys": "properties.gclid",
+    "required": false,
+    "metadata": {
+      "type": "toString"
+    }
+  },
+  {
+    "destKey": "conversionAdjustments[0].gclidDateTimePair.conversionDateTime",
+    "sourceKeys": "properties.conversionDateTime",
```
Comment (utsabc) [inline v0/destinations/google_ads_enhanced_conversion/data/trackConfig.json:17]: Lets rephrase this to `conversionTime`

File: v0/destinations/google_ads_enhanced_conversion/data/trackConfig.json [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+    "destKey": "conversionAdjustments[0].orderId",
+    "sourceKeys": "properties.orderId",
+    "required": true,
+    "metadata": {
+      "type": "toString"
+    }
+  },
+  {
+    "destKey": "conversionAdjustments[0].adjustmentDateTime",
+    "sourceKeys": "properties.adjustmentDateTime",
```
Comment (utsabc) [inline v0/destinations/google_ads_enhanced_conversion/data/trackConfig.json:46]:
```suggestion
    "sourceKeys": "properties.adjustmentTime",
```

File: v0/destinations/google_ads_enhanced_conversion/data/trackConfig.json [hunk 14]
```diff
@@ -0,0 +1,129 @@
...
+    "destKey": "conversionAdjustments[0].userAgent",
+    "sourceKeys": "context.userAgent",
+    "required": false,
+    "metadata": {
+      "type": "toString"
+    }
+  },
+  {
+    "destKey": "conversionAdjustments[0].userIdentifiers[0].hashedEmail",
+    "sourceKeys": ["context.traits.email"],
```
Comment (utsabc) [inline v0/destinations/google_ads_enhanced_conversion/data/trackConfig.json:59]: we should use `GenericFieldMapping.json` for this

File: v0/destinations/google_ads_enhanced_conversion/data/trackConfig.json [hunk 15]
```diff
@@ -0,0 +1,129 @@
...
+    "required": false,
+    "metadata": {
+      "type": "hashToSha256"
+    }
+  },
+  {
+    "destKey": "conversionAdjustments[0].userIdentifiers[1].hashedPhoneNumber",
+    "sourceKeys": ["context.traits.phone"],
```
Comment (utsabc) [inline v0/destinations/google_ads_enhanced_conversion/data/trackConfig.json:67]: Same here

File: v0/destinations/google_ads_enhanced_conversion/data/trackConfig.json [hunk 16]
```diff
@@ -0,0 +1,129 @@
...
+    "required": false,
+    "metadata": {
+      "type": "hashToSha256"
+    }
+  },
+  {
+    "destKey": "conversionAdjustments[0].userIdentifiers[2].addressInfo.hashedFirstName",
+    "sourceKeys": ["context.traits.firstName"],
```
Comment (utsabc) [inline v0/destinations/google_ads_enhanced_conversion/data/trackConfig.json:75]: Same here

File: v0/destinations/google_ads_enhanced_conversion/data/trackConfig.json [hunk 17]
```diff
@@ -0,0 +1,129 @@
...
+    "required": false,
+    "metadata": {
+      "type": "hashToSha256"
+    }
+  },
+  {
+    "destKey": "conversionAdjustments[0].userIdentifiers[2].addressInfo.hashedLastName",
+    "sourceKeys": ["context.traits.lastName"],
```
Comment (utsabc) [inline v0/destinations/google_ads_enhanced_conversion/data/trackConfig.json:83]: Same here

File: v0/destinations/google_ads_enhanced_conversion/data/trackConfig.json [hunk 18]
```diff
@@ -0,0 +1,129 @@
...
+    "required": false,
+    "metadata": {
+      "type": "hashToSha256"
+    }
+  },
+  {
+    "destKey": "conversionAdjustments[0].userIdentifiers[2].addressInfo.state",
+    "sourceKeys": "context.traits.state",
```
Comment (utsabc) [inline v0/destinations/google_ads_enhanced_conversion/data/trackConfig.json:91]: we should consider presence of an `address` object

File: v0/destinations/google_ads_enhanced_conversion/data/trackConfig.json [hunk 19]
```diff
@@ -0,0 +1,129 @@
+[
+  {
+    "destKey": "partialFailure",
+    "sourceKeys": "properties.partialFailure",
```
Comment (utsabc) [inline v0/destinations/google_ads_enhanced_conversion/data/trackConfig.json:4]: Google suggests to keep this `true` by default we shouldn't take as an input rather default it

File: v0/destinations/google_ads_enhanced_conversion/data/trackConfig.json (hunk 13 above)
Comment (utsabc) [reply v0/destinations/google_ads_enhanced_conversion/data/trackConfig.json:46]: Also we can take `timeStamp` if the above is not present as this considers the time of adjustment

File: v0/destinations/google_ads_enhanced_conversion/transform.js [hunk 20]
```diff
@@ -0,0 +1,122 @@
...
+  }
+  if (typeof payload.partialFailure !== "boolean") {
+    throw new CustomError(
+      `[Google_ads_enhanced_marketing]:: partialFailure should be a boolean.`,
+      400
+    );
+  }
+  payload.conversionAdjustments[0].adjustmentType = "ENHANCEMENT";
+  // Removing the null values from userIdentifier
+  const arr = payload.conversionAdjustments[0].userIdentifiers;
```
Comment (utsabc) [inline v0/destinations/google_ads_enhanced_conversion/transform.js:66]: As we are hashing the user identifiers would it be possible to hash null values? If not this check is unnecessary. Also if there is null value can we consider hashing empty string as user identifier?

File: v0/destinations/google_ads_enhanced_conversion/util.js [hunk 21]
```diff
@@ -0,0 +1,148 @@
...
+const { isHttpStatusSuccess } = require("../../util/index");
+const { TRANSFORMER_METRIC } = require("../../util/constant");
+const ErrorBuilder = require("../../util/error");
+const {
+  DISABLE_DEST,
+  REFRESH_TOKEN
+} = require("../../../adapters/networkhandler/authConstants");
+const Cache = require("../../util/cache");
+
+const conversionActionIdCache = new Cache(24 * 60 * 60);
```
Comment (utsabc) [inline v0/destinations/google_ads_enhanced_conversion/util.js:13]: We should store the cache time in constant and source it here

File: v0/destinations/google_ads_enhanced_conversion/util.js [hunk 22]
```diff
@@ -0,0 +1,148 @@
...
+
+/**
+ *  This function is used for collecting the conversionActionId using the conversion name
+ * @param {*} method
+ * @param {*} headers
+ * @param {*} params
+ * @returns
+ */
+
+const getConversionActionId = async (method, headers, params) => {
```
Comment (utsabc) [inline v0/destinations/google_ads_enhanced_conversion/util.js:28]: Did we test out this functionality? 
 - Maybe set a lower cache expiry and tets it out?

File: v0/destinations/google_ads_enhanced_conversion/util.js [hunk 23]
```diff
@@ -0,0 +1,148 @@
...
+    method,
+    headers,
+    params
+  );
+  if (
+    !conversionActionIdResponse.success &&
+    !isHttpStatusSuccess(conversionActionIdResponse.response.response.status)
+  ) {
+    // eslint-disable-next-line prefer-destructuring
+    set(
```
Comment (utsabc) [inline v0/destinations/google_ads_enhanced_conversion/util.js:68]: Why not return a new response rather than mutating the existing one, it lowers risk

File: v0/destinations/google_ads_enhanced_conversion/util.js [hunk 24]
```diff
@@ -0,0 +1,148 @@
...
+ * @returns
+ */
+
+const getConversionActionId = async (method, headers, params) => {
+  const conversionActionIdKey = sha256(
+    params.event + params.customerId
+  ).toString();
+  return conversionActionIdCache.get(conversionActionIdKey, async () => {
+    const data = {
+      query: `SELECT conversion_action.id FROM conversion_action WHERE conversion_action.name = '${params.event}'`
```
Comment (utsabc) [inline v0/destinations/google_ads_enhanced_conversion/util.js:34]: We should test some negative scenarios 
 - 4xx scenarios
 - Wrong query data
 - wring event name
 - wrong id

**Review by utsabc — CHANGES_REQUESTED** [review]

File: __mocks__/data/google_adwords_remarketing_lists/proxy_response.json (hunk 11 above)
Comment (ItsSudip) [reply __mocks__/data/google_adwords_remarketing_lists/proxy_response.json:84]: I think this suggestion occurred during the git hub explanation in the happy hour call. Can you resolve this?

File: v0/destinations/google_ads_enhanced_conversion/data/trackConfig.json (hunk 12 above)
Comment (ItsSudip) [reply v0/destinations/google_ads_enhanced_conversion/data/trackConfig.json:17]: I am using this name for parity with the destination name.

File: v0/destinations/google_ads_enhanced_conversion/data/trackConfig.json (hunk 13 above)
Comment (ItsSudip) [reply v0/destinations/google_ads_enhanced_conversion/data/trackConfig.json:46]: Our timestamp format is not same as google's. Google's format is `The format is "yyyy-mm-dd hh:mm:ss+|-hh:mm", e.g. "2019-01-01 12:32:45-08:00"` and our format is `2019-01-01T11:15:18.299Z`

File: v0/destinations/google_ads_enhanced_conversion/transform.js (hunk 20 above)
Comment (ItsSudip) [reply v0/destinations/google_ads_enhanced_conversion/transform.js:66]: This check is basically creating the userIdentifier object. If you check in trackConfig.json, there I am mapping every userIdentifiers position-wise. So if some of the userIdentifier is not exist then I am putting a null value. After that, I am removing all null values from the array.

File: v0/destinations/google_ads_enhanced_conversion/util.js (hunk 21 above)
Comment (ItsSudip) [reply v0/destinations/google_ads_enhanced_conversion/util.js:13]: Made that a constant and provided scope for getting from .env also.

File: v0/destinations/google_ads_enhanced_conversion/util.js (hunk 23 above)
Comment (ItsSudip) [reply v0/destinations/google_ads_enhanced_conversion/util.js:68]: Here, in enhanced conversion, google is returning the data as an array but for remarketing list, it was an object. So mutating it.

File: v0/destinations/google_ads_enhanced_conversion/util.js (hunk 22 above)
Comment (ItsSudip) [reply v0/destinations/google_ads_enhanced_conversion/util.js:28]: No

File: v0/destinations/google_ads_enhanced_conversion/util.js (hunk 24 above)
Comment (ItsSudip) [reply v0/destinations/google_ads_enhanced_conversion/util.js:34]: Apart from wrong query data I have tested others. I don't think our customer can change the query data.

Comment (utsabc) [reply v0/destinations/google_ads_enhanced_conversion/util.js:34]: Got it

File: v0/destinations/google_ads_enhanced_conversion/util.js (hunk 23 above)
Comment (ItsSudip) [reply v0/destinations/google_ads_enhanced_conversion/util.js]: Updated this.

**Review by utsabc — DISMISSED** [review]

File: v0/destinations/google_ads_enhanced_conversion/data/trackConfig.json (hunk 13 above)
Comment (ItsSudip) [reply v0/destinations/google_ads_enhanced_conversion/data/trackConfig.json]: I tried to use the originalTimestamp with the required format but the API is throwing `LATER_THAN_MAXIMUM_DATE` this error.

Comment (utsabc) [reply v0/destinations/google_ads_enhanced_conversion/data/trackConfig.json]: Ok lets keep the same

**Review by utsabc — APPROVED** [review]

**Review by saikumarrs — APPROVED** [review]

**Review by arajguha — APPROVED** [review]
